File: backend/src/services/authService.js
import bcrypt from "bcryptjs";
import jwt from "jsonwebtoken";
import { v4 as uuidv4 } from "uuid";
import config from "../config.js";
import { createUser, getUserByEmail } from "../infrastructure/dynamodb.js";

export async function signup(email, password) {
  const existing = await getUserByEmail(email);
  if (existing) {
    const err = new Error("User with this email already exists");
    err.status = 409;
    throw err;
  }

  const passwordHash = await bcrypt.hash(password, 10);
  const user = {
    user_id: uuidv4(),
    email,
    password_hash: passwordHash,
    created_at: new Date().toISOString(),
  };

  await createUser(user);

  const token = generateToken(user);
  return { token, user_id: user.user_id, email: user.email };
}

export async function login(email, password) {
  const user = await getUserByEmail(email);
  if (!user) {
    const err = new Error("Invalid email or password");
    err.status = 401;
    throw err;
  }

  const valid = await bcrypt.compare(password, user.password_hash);
  if (!valid) {
    const err = new Error("Invalid email or password");
    err.status = 401;
    throw err;
  }

  const token = generateToken(user);
  return { token, user_id: user.user_id, email: user.email };
}

export function verifyToken(token) {
  try {
    return jwt.verify(token, config.jwt.secret);
  } catch {
    const err = new Error("Invalid or expired token");
    err.status = 401;
    throw err;
  }
}

function generateToken(user) {
  return jwt.sign(
    { sub: user.user_id, email: user.email },
    config.jwt.secret,
    { expiresIn: config.jwt.expiresIn }
  );
}

File: backend/app/services/auth_service.py
import logging
import uuid
from datetime import datetime, timedelta

import bcrypt
from jose import JWTError, jwt

from app.config import Settings
from app.domain.entities import AuthResponse, User
from app.domain.exceptions import AuthenticationError, DuplicateError
from app.infrastructure.dynamodb.user_repo import UserRepository

logger = logging.getLogger(__name__)


class AuthService:
    def __init__(self, user_repo: UserRepository, settings: Settings):
        self._user_repo = user_repo
        self._settings = settings

    def signup(self, email: str, password: str) -> AuthResponse:
        existing = self._user_repo.get_by_email(email)
        if existing:
            raise DuplicateError("User with this email already exists")

        password_hash = bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode()
        user = User(
            user_id=str(uuid.uuid4()),
            email=email,
            password_hash=password_hash,
        )
        self._user_repo.create(user)

        token = self._create_token(user)
        logger.info("User signed up: %s", user.user_id)
        return AuthResponse(token=token, user_id=user.user_id, email=user.email)

    def login(self, email: str, password: str) -> AuthResponse:
        user = self._user_repo.get_by_email(email)
        if not user:
            raise AuthenticationError("Invalid email or password")

        if not bcrypt.checkpw(password.encode(), user.password_hash.encode()):
            raise AuthenticationError("Invalid email or password")

        token = self._create_token(user)
        logger.info("User logged in: %s", user.user_id)
        return AuthResponse(token=token, user_id=user.user_id, email=user.email)

    def verify_token(self, token: str) -> str:
        try:
            payload = jwt.decode(
                token,
                self._settings.JWT_SECRET,
                algorithms=[self._settings.JWT_ALGORITHM],
            )
            user_id: str = payload.get("sub")
            if user_id is None:
                raise AuthenticationError("Invalid token")
            return user_id
        except JWTError as e:
            raise AuthenticationError(f"Invalid token: {e}")

    def _create_token(self, user: User) -> str:
        payload = {
            "sub": user.user_id,
            "email": user.email,
            "iat": datetime.utcnow(),
            "exp": datetime.utcnow() + timedelta(hours=self._settings.JWT_EXPIRY_HOURS),
        }
        return jwt.encode(
            payload,
            self._settings.JWT_SECRET,
            algorithm=self._settings.JWT_ALGORITHM,
        )
